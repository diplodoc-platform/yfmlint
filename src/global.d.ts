declare module 'markdownlint' {
    export * from 'markdownlint/lib/exports.mjs';
}

declare module 'markdownlint/promise' {
    export * from 'markdownlint/lib/exports-promise.mjs';
}
