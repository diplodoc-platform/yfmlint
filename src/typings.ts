import type {Dictionary} from 'lodash';
// eslint-disable-next-line @typescript-eslint/no-redeclare
import type {Plugin} from 'markdownlint/lib/markdownlint';
import type {LogLevels} from './utils';

export interface LintConfig {
    default?: boolean;
    'log-levels': Record<string, LogLevels>;
    [x: string]: unknown;
}

export interface Options {
    plugins?: Function[] | Plugin;
    pluginOptions: Record<string, unknown>;
    lintConfig?: LintConfig;
    sourceMap?: Dictionary<string>;
    frontMatter?: RegExp | null;
}

export interface Logger {
    info(...args: string[]): void;
    warn(...args: string[]): void;
    error(...args: string[]): void;
}
