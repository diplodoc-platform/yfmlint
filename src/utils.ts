import type {LintConfig, LintError, Logger, RawLintConfig} from './typings';

import merge from 'lodash/merge';

export enum LogLevels {
    INFO = 'info',
    WARN = 'warn',
    ERROR = 'error',
    DISABLED = 'disabled',
}

const DEFAULT_LOG_LEVEL = LogLevels.WARN;

export function normalizeConfig(...parts: RawLintConfig[]) {
    const normalized = parts.map((part) => {
        return Object.keys(part).reduce((result, key) => {
            if (key === 'default') {
                result[key] = part[key];
            } else {
                if (typeof part[key] === 'boolean') {
                    result[key] = part[key]
                        ? {
                              level: DEFAULT_LOG_LEVEL,
                          }
                        : false;
                }

                if (typeof part[key] === 'string') {
                    result[key] = {
                        level: part[key] as LogLevels,
                    };
                }

                if (part[key] && typeof part[key] === 'object') {
                    result[key] = {
                        ...(part[key] as object),
                        // @ts-ignore
                        level: part[key].level || DEFAULT_LOG_LEVEL,
                    };
                }
            }

            return result;
        }, {} as LintConfig);
    });

    return merge({}, ...normalized);
}

export function getLogLevel(logLevels: LintConfig, ruleNames: string[]) {
    return (
        ruleNames.map((ruleName) => {
            if (!logLevels[ruleName]) {
                return LogLevels.DISABLED;
            }

            // @ts-ignore
            return logLevels[ruleName].level;
        })[0] || DEFAULT_LOG_LEVEL
    );
}

export function log(errors: LintError[], logger: Logger) {
    for (const error of errors) {
        const message = String(error);

        switch (error.level) {
            case LogLevels.ERROR:
                logger.error(message);
                break;
            case LogLevels.WARN:
                logger.warn(message);
                break;
            case LogLevels.DISABLED:
            default:
                break;
        }
    }
}
