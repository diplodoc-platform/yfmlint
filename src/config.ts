import {LogLevels} from './utils';

export default {
    // Default state for all rules
    default: false,

    // Inline code length
    YFM001: {
        level: LogLevels.WARN,
        maximum: 100,
    },
    YFM002: LogLevels.WARN, // No header found in the file for the link text
    YFM003: LogLevels.ERROR, // Link is unreachable
    YFM004: LogLevels.ERROR, // Table not closed
    YFM005: LogLevels.ERROR, // Tab list not closed
    YFM006: LogLevels.WARN, // Term definition duplicated
    YFM007: LogLevels.WARN, // Term used without definition
    YFM008: LogLevels.WARN, // Term inside definition not allowed
    YFM009: LogLevels.WARN, // Term definition used not at the end of file
    YFM010: LogLevels.WARN, // Autotitle anchor is missed
};
