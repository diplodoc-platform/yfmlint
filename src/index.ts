import type {LintError, Options} from './typings';

import merge from 'lodash/merge';
import attrs from 'markdown-it-attrs';

import defaultLintConfig from './config';
import * as rules from './rules';
import {LogLevels, getLogLevel} from './utils';

export async function yfmlint(content: string, path: string, opts: Options) {
    // @ts-ignore
    const {lint} = import('markdownlint/promise');
    const {plugins: customPlugins, pluginOptions, frontMatter = null} = opts;

    pluginOptions.isLintRun = true;

    const config = merge({}, defaultLintConfig, opts.lintConfig || {});

    const plugins = customPlugins && [attrs, ...customPlugins];
    const preparedPlugins = plugins && plugins.map((plugin) => [plugin, pluginOptions]);

    try {
        const errors = await lint({
            strings: {[path]: content},
            markdownItPlugins: preparedPlugins,
            handleRuleFailures: true,
            frontMatter,
            config,
            customRules: Object.values(rules),
        })[path];

        return errors.map((error: LintError) => {
            error.level = getLogLevel(config['log-levels'], error.ruleNames, LogLevels.WARN);
            error.toString = function () {
                const {
                    lineNumber,
                    errorDetail,
                    errorContext,
                    ruleNames,
                    // @ts-expect-error bad markdownlint typings
                    ruleName,
                    // @ts-expect-error bad markdownlint typings
                    ruleAlias,
                    ruleDescription,
                } = error;
                const ruleMoniker = ruleNames
                    ? ruleNames.join(' / ')
                    : ruleName + ' / ' + ruleAlias;

                return (
                    `${path}${lineNumber ? `: ${lineNumber}:` : ':'} ${ruleMoniker} ${ruleDescription}` +
                    (errorDetail ? ` [${errorDetail}]` : '') +
                    (errorContext ? ` [Context: "${errorContext}"]` : '')
                );
            };

            return error;
        });
    } catch {}
}
