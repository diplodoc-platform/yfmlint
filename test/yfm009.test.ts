import {describe, expect, it} from 'vitest';
import term from '@diplodoc/transform/lib/plugins/term';
import dedent from 'ts-dedent';

import {LogLevels, yfmlint} from '../src';

import {formatErrors} from './utils';

const withIncludes = dedent`
    [*widget-popup1]: {% include [ ](./_includes/widget.md) %}

    [*button-popup2]: {% include [ ](./_includes/button) %}

    [*button-popup3]: {% include [ ](./_includes/button) %}
`;

const plainText = dedent`
    [*widget-popup1]: hello world

    hi i am text

    [*widget-popup2]: it's will fail
`;

const appendText = (text: string) => {
    return `${text}\n\nhi i am text after`;
};

const prependText = (text: string) => {
    return `hi i am text before\n\n${text}`;
};

const lint = (input: string, path: string) => {
    return yfmlint(input, path, {
        lintConfig: {
            YFM009: LogLevels.ERROR,
        },
        plugins: [term],
    });
};

describe('YFM009', () => {
    it('not accepts text between terms', async () => {
        const errors = (await lint(plainText, 'test1.md')) || [];
        expect(formatErrors(errors)).toMatchSnapshot();
    });

    it('not accepts includes with text after', async () => {
        const errors = (await lint(appendText(withIncludes), 'test2.md')) || [];
        expect(formatErrors(errors)).toMatchSnapshot();
    });

    it('accepts includes', async () => {
        const errors = (await lint(withIncludes, 'test3.md')) || [];
        expect(formatErrors(errors)).toMatchSnapshot();
    });

    it('accepts includes with text before', async () => {
        const errors = (await lint(prependText(withIncludes), 'test4.md')) || [];
        expect(formatErrors(errors)).toMatchSnapshot();
    });
});
