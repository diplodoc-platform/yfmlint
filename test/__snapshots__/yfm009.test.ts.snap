// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`YFM009 > accepts includes 1`] = `[]`;

exports[`YFM009 > accepts includes with text before 1`] = `[]`;

exports[`YFM009 > not accepts includes with text after 1`] = `
[
  "test2.md: 7: YFM009 / no-term-definition-in-content Term definition should be placed at the end of file [There is a content between term definition. All term defitions should be placed at the end of file.]",
  "test2.md: 7: YFM009 / no-term-definition-in-content Term definition should be placed at the end of file [The file must end with term only.]",
]
`;

exports[`YFM009 > not accepts text between terms 1`] = `
[
  "test1.md: 3: YFM009 / no-term-definition-in-content Term definition should be placed at the end of file [There is a content between term definition. All term defitions should be placed at the end of file.]",
]
`;
